LFS-663/LFS-973: The dialog for editing questions should only display question properties and options that are relevant for the selected data type

marta- opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-973l)

- [x] Some data types have a predefined display mode (e.g. boolean - list of options; date, time, computed - input; file - file uploader) and selecting another has no effect on how the question is displayed. The `displayMode` is now shown only when it is indeed configurable.
- [x] The user can now enter Answer options  only if `displayMode` is `list` or `list+input`
- [x] Some data types (date, time, computed) currently support at most one answer. The wizard now prevents the user from selecting other restrictions for the number of answers.
- [x] The display modes `none` and `hidden` are currently not supported and have been (temporarily) removed from the wizard UI.
- [x] Also displayed helper text explaining that for `maxAnswers` and `maxPerSubject` 0 means Unlimited.

marta- committed:
LFS-663/LFS-973: The dialog for editing questions should only display question properties and options that are relevant for the selected data type

marta- committed:
[misc] Added helper text to maxAnswers and maxPerSubject informing the user that value 0 means "Unlimited"

sashaandjic commented:
tested - looks good!
![image](https://user-images.githubusercontent.com/23243244/111550347-4df77e00-8754-11eb-8fd2-5845fa704b56.png)

sashaandjic commented:
Was the `Vocabulary filter` added in this PR as well? It is not functional.

marta- commented:
> Was the `Vocabulary filter` added in this PR as well? It is not functional.

No, it's still the old one that doesn't work. It will be done separately, there's a task for it: https://phenotips.atlassian.net/browse/LFS-884

marta- committed:
LFS-663/LFS-973: The dialog for editing questions should only display question properties and options that are relevant for the selected data type

Addressing comments: also removed the (currently non-functional) `vocabularyFilter` property for vocabulary questions.

sashaandjic reviewed (approved):
retested - Vocabulary filter option is removed
![image](https://user-images.githubusercontent.com/23243244/111552381-96189f80-8758-11eb-9c8c-97d3f577e63b.png)

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -72,8 +103,7 @@
         "upperLimit": "string",
-        "minAnswers": "long",
-        "maxAnswers": "long",
+        "minAnswers": {"0":{}, "1":{}},
```
Just trying to get the idea here. Why this change was made?

marta- replied:
The current code of the Time Question is designed around the assumption that `maxAnswers = 1`, therefore:
* The option to configure`maxAnswers` is being removed from the wizard UI as it would have no effect
* The only values that make sense for `minAnswers` are `0` and `1`, and this syntax is the easiest way to produce a drop-down list.

Boolean, Date and Pedigree questions have a similar setup.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/NumberInput.jsx:
```diff
@@ -35,6 +35,7 @@ import QuestionComponentManager from "./QuestionComponentManager";
 let NumberInput = (props) => {
   let { objectKey, data } = props;
   const defaultValue = props.value === "long" ? 0 : '';
+  const isMax = props.value === "long" && objectKey.startsWith('max');
```
Why not just check for `maxAnswers`? I don't want to hardcode the idea that any property starting with `max` has `0=unlimited`.

marta- replied:
This is for both `maxAnswers` and `maxPerSubject`, and introducing the **convention** that for any `Long` property that starts with `max` 0 means Unlimited was exactly my intention. Note that `maxValue` is excluded because it's not `Long`.

sdumitriu replied:
OK, let's hope it won't byte us in the future.
